### DISCUSSION TITLE
generic error category template review

### DISCUSSION THREAD

**Discussion by arturbac** [arturbac/simple_enum#1] [Q&A] [answered]
@stephenberry Can You take a look when You will have time at generic_error_category template for use with expected/std::error_code as error category template. I am interested what would You change or improve.
The base idea is to remove from users copy/pasting code, and making it easy to create a error category with just 1 explicit instantiation.

https://arturbac.github.io/simple_enum/generic_error_category.html
https://github.com/arturbac/simple_enum/blob/master/examples/generic_error_category.cc

**Comment by stephenberry** [comment, marked as answer]
This is cool. But, I'm a little confused as to why you need a strongly typed `generic_error_category` when it seems in the end you are only printing the category name.

Can you just include the name as a field of a specialization for your enum? Like below:

```c++
template <>
struct simple_enum::meta<custom_error_example::function_error>
{
   using enum function_error;
   static constexpr simple_enum::adl_info bounds{failed_other_reason, unhandled_exaption};

   static constexpr std::string_view category = "Function Error Category Name";
};
```

If you can explain why the above is not desirable I think I'll better understand your goals.

**Reply by arturbac** [reply]
You mean by passing additional category = "Function Error Category Name" ?
Just to handle custom Category name at compile time, I am thinking of taking just enumeration name and do Camel space Casing as with enum name values.

And yes, below concept below would make sense.
```
template <>
struct simple_enum::meta<custom_error_example::function_error>
{
   using enum function_error;
   static constexpr std::string_view category = "Function Error Category Name";
};
```
simple_enum::adl_info enum bounds are unrelated as a way to specify bound by adl., there is also and class spcialization with info
[bounds](https://github.com/arturbac/simple_enum/blob/master/docs/enum_bounds.md) 
If I require puting category there then I will narrow options to specify bounds.

But I think I will go with
- by default use enumeration name Camel space Cased so enum my_error_enum will give My Error Enum category name
- add option like the idea with meta You proposed to customise the name.

Thx for Your review.

**Reply by stephenberry** [reply]
Awesome! Thanks for sharing this approach/feature.

**Reply by arturbac** [reply]
You are welcome, I am just writing practical things I need at production for myself in few projects at my corpo, the same as You do with awesome `glaze`, that I use in my projects.
I think one of my nicest use of `glaze` I did is universal_config I wrote, because I have plenty of configuration files and do not want to waste time on writing same code from scratch.
Maybe it is some idea for such functionality in glaze ?
[header](https://github.com/arturbac/kdevcxx_with_ai/blob/master/ai_processing/include/aiprocess/universal_config.h) [impl](https://github.com/arturbac/kdevcxx_with_ai/blob/master/ai_processing/include/aiprocess/universal_config.hpp)

**Reply by stephenberry** [reply]
That is very useful. I'll consider this approach for my own configuration files and adding functionality to glaze. Your code looks great.

**Reply by arturbac** [reply]
I found a nice way utilize something You have to do anyway std::is_error_code_enum spec..

```
template<typename ErrorEnum>
concept error_category_name_specialized = requires {
  { std::is_error_code_enum<ErrorEnum>::category_name } -> std::same_as<std::string_view const &>;
    
template<>
struct std::is_error_code_enum<test_error_class_spec> : true_type
  {
  static constexpr std::string_view category_name = "My Error Category";
  };
```
So there will be default from type name and with ability to override with custom from std::is_error_code_enum

**Reply by stephenberry** [reply]
Really nice!

**Reply by arturbac** [reply]
Change is done and released.
https://arturbac.github.io/simple_enum/generic_error_category.html

**Reply by arturbac** [reply]
I just got idea to simplify it more
I will provide partial specialization of std::is_error_code_enum for 
enums for which there is adl function, so It will allow avoiding annying global namespace std:: declarations and just declare this function in the enumeration namespace

```
namespace user_namespace
{
enum class enumeration_type{};
consteval auto adl_decl_error_code(enumeration_type) -> bool { return true;}
}
```
And in simple enum there will be partial spec
```
template<typename T>
  requires simple_enum::concepts::declared_error_code<T>
struct std::is_error_code_enum<T> : std::true_type
  {
  };
```

**Reply by arturbac** [reply]
A now I am using it at corpo with glaze error_code

![obraz](https://github.com/arturbac/simple_enum/assets/14975842/d006ae1a-f136-4d78-99d2-b46398412806)
![obraz](https://github.com/arturbac/simple_enum/assets/14975842/6f4c5cd1-7005-448a-9386-a7a2c7da020e)
